File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/fs/http/HttpInputStream.java
package app.zxtune.fs.http;

import java.io.IOException;
import java.io.InputStream;
import java.net.HttpURLConnection;

final class HttpInputStream extends InputStream {

  private HttpURLConnection connection;
  private InputStream delegate;

  HttpInputStream(HttpURLConnection connection) throws IOException {
    this.connection = connection;
    this.delegate = connection.getInputStream();
  }

  @Override
  public void close() {
    delegate = null;
    if (connection != null) {
      connection.disconnect();
      connection = null;
    }
  }

  @Override
  public int available() throws IOException {
    return delegate.available();
  }

  @Override
  public int read() throws IOException {
    return delegate.read();
  }

  @Override
  public int read(byte[] b, int off, int len) throws IOException {
    return delegate.read(b, off, len);
  }

  @Override
  public int read(byte[] b) throws IOException {
    return delegate.read(b);
  }

  @Override
  public long skip(long n) throws IOException {
    return delegate.skip(n);
  }
}

File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/fs/http/HttpInputStream.kt
package app.zxtune.fs.http

import java.io.InputStream
import java.net.HttpURLConnection
import java.util.zip.GZIPInputStream

internal class HttpInputStream(connection: HttpURLConnection) : InputStream() {

    private var connection: HttpURLConnection? = connection

    // Just for tests - no transparent compression...
    // https://developer.android.com/reference/java/net/HttpURLConnection.html#performance :
    // By default, this implementation of HttpURLConnection requests that servers use gzip
    // compression and it automatically decompresses the data for callers of
    // URLConnection.getInputStream(). The Content-Encoding and Content-Length response headers
    // are cleared in this case.
    private var delegate: InputStream? = if ("gzip" == connection.contentEncoding) {
        GZIPInputStream(connection.inputStream)
    } else {
        connection.inputStream
    }

    override fun close() {
        delegate = null
        connection?.disconnect()
        connection = null
    }

    override fun available() = delegate!!.available()

    override fun read() = delegate!!.read()

    override fun read(b: ByteArray, off: Int, len: Int) = delegate!!.read(b, off, len)

    override fun read(b: ByteArray) = delegate!!.read(b)

    override fun skip(n: Long) = delegate!!.skip(n)
}
